README.md
# HP Operations Orchestration client

This is a python library and command line tool to interact with HP OO. The motivation is that you can use the command line tool as part of a continuous integration pipeline. Unlike the HP tool it has sensible exit codes.

## Install

I recommend you use a virtual environment. See [here](http://virtualenvwrapper.readthedocs.org/en/latest/install.html) for instructions on virtualenvwrapper. Then `mkvirtualenv oo_client` followed by:

```
pip install git+https://bitbucket.org/automationlogic/oo_client.git
```

Now run `hpoo -h` to see the help for the command line tool.

Only dependencies are the python requests module which pip will install for you. If you want to use the build function you also need a JDK (for the jar command line tool). Look for a java-devel package on your platform.

## Build

Currently this is the least flexible part of the tool, it will build a content pack from an SVN checkout and optionally create release branches and tags. A simple example:

```
svn co svn://localhost/srv/svn/mmb-oo
cd mmb-oo/trunk/test-content/
mkdir target
hpoo -a build -bp ./
```

Example output:
```
INFO:ContentBuilder:Creating content pack test-content-cp-2.0.1-109-SNAPSHOT.jar
INFO:ContentBuilder:Done
```

## Deploy

This allows you to deploy one or more content packs to a central, if you specify multiple content packs OO will deploy them in the correct order for dependencies.

```
hpoo -a deploy -cp ./target/test-content-cp-2.0.1-109-SNAPSHOT.jar -c https://central.local:8443 -u admin -p admin
```

Example output:
```
INFO:OOClient:Got new deployment ID: 169300006
INFO:OOClient:Starting uploads...
INFO:OOClient:Uploaded test-content-cp-2.0.1-109-SNAPSHOT.jar
INFO:OOClient:Deployment started, waiting up to 300s to complete
INFO:OOClient:Deployed test-content-cp-2.0.1-109-SNAPSHOT.jar successfully
```

## Test

This mode will run all the flows matching a certain pattern and exit 0 only if all the flows succeed. Example:

```
hpoo -a integration_test -cp test-content --filter tests/integration_tests/test_ -c https://central.local:8443 -u admin -p admin
```

Example output:
```
INFO:IntegrationTester:Found the following test flows, running sequentially
INFO:IntegrationTester:Library/mmb1/tests/integration_tests/test_mmb-test.xml
INFO:IntegrationTester:Library/mmb1/tests/integration_tests/test_mmb-test_fail.xml
INFO:OOClient:Running flow: Library/mmb1/tests/integration_tests/test_mmb-test.xml
INFO:OOClient:Flow Library/mmb1/tests/integration_tests/test_mmb-test.xml FINISHED: RESOLVED, link https://central.local:8445/oo/#/runtimeWorkspace/runs/167900122
INFO:OOClient:Running flow: Library/mmb1/tests/integration_tests/test_mmb-test_fail.xml
INFO:OOClient:Flow Library/mmb1/tests/integration_tests/test_mmb-test_fail.xml FINISHED: RESOLVED, link https://central.local:8445/oo/#/runtimeWorkspace/runs/167900135
```

Currently the flows run sequentially but we could make them run in parallel.

## Run flow

You can also run a flow by path or uuid:

```
hpoo -a run -f Library/mmb1/tests/integration_tests/test_mmb-test.xml -c https://central.local:8445 -u admin -p admin
```

## Develop

To run the tests:
```
git clone git@bitbucket.org:automationlogic/oo_client.git
python setup.py test
```

Example using the library yourself:
```
import oo_client.hpoo
c = oo_client.hpoo.OOClient("https://localhost:8443", "admin", "pass", ssl=False)
c.get_run_status(167900135)
```
or to make any rest call that doesn't have a helper method:
```
c.rest.get('version')
```
## Config

This mode can get/set configuration item(s) with four options.

Example setting the value of a system-properties with the specified path to 'blah':

```
hpoo -k -a set_config -path sp1 -type system-properties -value blah -c https://central.local:8445 -u admin -p cloud
```
This is to get a system-properties specified in 'sp1':
```
hpoo -k -a get_config -path sp1 -type system-properties -c https://central.local:8445 -u admin -p cloud
```

When no path but type is specified, get_config will return all configuration items with the specified type:
```
hpoo -k -a set_config -type system-properties -value blah -c https://central.local:8445 -u admin -p cloud
```
While neither is specified, it will return all configuration items:
```
hpoo -k -a set_config -path sp1 -type system-properties -value blah -c https://central.local:8445 -u admin -p cloud
```

CHANGELOG.md
#0.5

Add support for getting/setting configuration items in central.

#0.4

Initial release


tests/test_hpoo_rest.py
import unittest

from mock import Mock

from oo_client.hpoo import OORestCaller

import oo_client.errors as errors

import requests


class TestHPOORest(unittest.TestCase):
    def setUp(self):
        self.mock_reqs = Mock()
        requests.Session = self.mock_reqs

    def test_post(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 200
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'post.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        assert rest.session.auth == ("aa", "bb")
        assert rest.session.verify is True
        ret = rest.post('some-path')
        url = "https://blah:1234/oo/rest/v1/some-path"
        mock_session.post.assert_called_with(url, None)
        self.assertEquals(ret, {'aaa': 'aaa'})

    def test_post_file(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 200
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'post.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        assert rest.session.auth == ("aa", "bb")
        assert rest.session.verify is True
        ret = rest.post('some-path', files=['some_file', 'some_other_file'])
        url = "https://blah:1234/oo/rest/v1/some-path"
        mock_session.post.assert_called_with(url,
                                             files=['some_file',
                                                    'some_other_file'])
        self.assertEquals(ret, {'aaa': 'aaa'})

    def test_post_json(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 200
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'post.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        assert rest.session.auth == ("aa", "bb")
        assert rest.session.verify is True
        ret = rest.post('some-path', data={'some': 'data'})
        url = "https://blah:1234/oo/rest/v1/some-path"
        mock_session.post.assert_called_with(url,
                                             '{"some": "data"}')
        self.assertEquals(ret, {'aaa': 'aaa'})

    def test_post_error(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 400
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'post.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        assert rest.session.auth == ("aa", "bb")
        assert rest.session.verify is True
        with self.assertRaises(errors.HTTPNon200):
            rest.post('some-path', Mock())

    def test_put(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 200
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'put.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        assert rest.session.auth == ("aa", "bb")
        assert rest.session.verify is True
        mock_data = Mock()
        rest.put('some-path', mock_data)
        url = "https://blah:1234/oo/rest/v1/some-path"
        mock_session.put.assert_called_with(url, mock_data)

    def test_put_error(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 400
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'put.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        with self.assertRaises(errors.HTTPNon200):
            rest.put('some-path', Mock())

    def test_get(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 200
        mock_response.text = '{"aaa": "aaa"}'
        mock_response.headers = {}
        mock_config = {'get.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        assert rest.session.auth == ("aa", "bb")
        assert rest.session.verify is True
        ret = rest.get('some-path')
        self.assertEquals(ret, {'aaa': 'aaa'})
        url = "https://blah:1234/oo/rest/v1/some-path"
        mock_session.get.assert_called_with(url, params={})

    def test_get_error(self):
        mock_session = Mock()
        mock_response = Mock()
        mock_response.status_code = 400
        mock_response.text = '{"aaa": "aaa"}'
        mock_config = {'get.return_value': mock_response}
        mock_session.configure_mock(**mock_config)
        self.mock_reqs.return_value = mock_session
        rest = OORestCaller("https://blah:1234", "aa", "bb")
        with self.assertRaises(errors.HTTPNon200):
            rest.get('some-path')

    def tearDown(self):
        pass


tests/test_hpoo_tester.py
import unittest

from mock import Mock, patch

from oo_client.hpoo_tester import IntegrationTester
import oo_client.errors as errors


class TestHPOOTester(unittest.TestCase):
    def setUp(self):
        pass

    def test_filter_flows(self):
        mock_client = Mock()
        it = IntegrationTester(mock_client, 'path/to_tests/test_')
        ret = it.filter_flows(['some/path/to/flow', 'another/path/to/flow'],
                              'some')
        self.assertEqual(ret, ['some/path/to/flow'])

    @patch('oo_client.hpoo_tester.IntegrationTester.filter_flows')
    def test_run_tests(self, mock_filter):
        mock_client = Mock()
        it = IntegrationTester(mock_client, 'some')
        flows = {123: 'some/path/to/flow', 456: 'another/path/to/flow'}
        mock_filter.return_value = ['some/path/to/flow']
        mock_client.get_all_flows_in_cp.return_value = flows
        mock_client.run_flows.return_value = True
        ret = it.run_tests('my-cp')
        mock_client.get_all_flows_in_cp.assert_called_with('my-cp')
        mock_filter.assert_called_with(['another/path/to/flow',
                                        'some/path/to/flow'],
                                       'some')
        mock_client.run_flows.assert_called_with(['some/path/to/flow'],
                                                 timeout=300)
        self.assertTrue(ret)
        with self.assertRaises(errors.NoFlowsFound):
            mock_filter.return_value = []
            ret = it.run_tests('my-cp')

    def tearDown(self):
        pass


oo_client/utils.py
import time
from uuid import UUID
import oo_client.errors as errors


def timeout(timeout, interval):
    def decorate(func):
        def wrapper(*args, **kwargs):
            attempts = 0
            while True:
                result = func(*args, **kwargs)
                if result:
                    return result
                if attempts >= timeout / interval:
                    raise errors.TimeoutError(format(func.__name__), timeout)
                attempts += 1
                time.sleep(interval)
        return wrapper
    return decorate


def validate_uuid4(uuid_string):
    """
    Validate that a UUID string is in
    fact a valid uuid4.
    Happily, the uuid module does the actual
    checking for us.
    It is vital that the 'version' kwarg be passed
    to the UUID() call, otherwise any 32-character
    hex string is considered valid.
    """
    try:
        val = UUID(uuid_string, version=4)
    except ValueError:
        # If it's a value error, then the string
        # is not a valid hex code for a UUID.
        return False
    # If the uuid_string is a valid hex code,
    # but an invalid uuid4,
    # the UUID.__init__ will convert it to a
    # valid uuid4. This is bad for validation purposes.
    return val.hex == uuid_string or val.__str__() == uuid_string


oo_client/errors.py
import logging
import json


class OOClientError(Exception):
    def __init__(self, msg):
        self.log = logging.getLogger(self.__class__.__name__)
        self.msg = msg
        self.log.error(msg)


class HTTPNon200(OOClientError):
    def __init__(self, status_code, response):
        self.msg = '{0}: {1}'.format(status_code, response)
        self.status_code = status_code
        try:
            self.response = json.loads(response)
        except Exception:
            self.response = response
        OOClientError(self.msg)


class ReleaseAlreadyExists(OOClientError):
    def __init__(self, version):
        self.msg = 'Version {0} has already been released, please pick a'\
                   'different version number'.format(version)
        self.version = version
        OOClientError(self.msg)


class NoFlowsFound(OOClientError):
    def __init__(self, filter):
        self.msg = 'No flows found with filter {0}'.format(filter)
        OOClientError(self.msg)


class TimeoutError(OOClientError):
    def __init__(self, function, time):
        self.msg = 'Timed out after waiting {0}s in {1}'.format(time, function)
        OOClientError(self.msg)


class NotFound(OOClientError):
    pass


oo_client/hpoo_tester.py
import logging
import oo_client.errors as errors


class IntegrationTester(object):
    def __init__(self, oo_client, test_filter):
        self.oo = oo_client
        self.int_test_path = test_filter
        self.log = logging.getLogger(self.__class__.__name__)

    def filter_flows(self, flows, path_filter):
        ret = [flow for flow in flows if path_filter in flow]
        return ret

    def run_tests(self, content_pack, timeout=300):
        flows = self.oo.get_all_flows_in_cp(content_pack)
        test_flows = self.filter_flows(flows.values(), self.int_test_path)
        if not test_flows:
            raise errors.NoFlowsFound(self.int_test_path)
        self.log.info("Found the following test flows, running sequentially")
        for flow in test_flows:
            self.log.info(flow)
        return self.oo.run_flows(test_flows, timeout=timeout)


oo_client/hpoo_builder.py
import StringIO
import xml.etree.ElementTree as ET
import ConfigParser
import os
import oo_client.errors as errors
from subprocess import check_call
from subprocess import call
from subprocess import check_output
import logging


class ContentBuilder(object):
    def __init__(self, path_to_content, svn_path=""):
        self.path = path_to_content
        self.config = self.read_props()
        self.cp_name = self.config.get('cp', 'content.pack.name')
        self.log = logging.getLogger(self.__class__.__name__)
        self.svn_path = svn_path

    def run_build(self, release=False, version=None, system_props=True,
                  system_accts=True, system_accts_passwds=False,
                  branch="trunk"):
        os.chdir(self.path)
        if release:
            if not version:
                version = self.config.get('cp', 'content.pack.version')
                version = version.replace('-SNAPSHOT', '')
            self.check_version_doesnt_exist(version)
            cp_name = '{0}-cp-{1}.jar'.format(self.cp_name, version)
            self.set_cp_version(version)
            if branch == 'trunk':
                self.log.info("Creating release branch")
                check_call(['svn', 'copy', '^{0}/trunk'.format(self.svn_path),
                            '^{0}/branches/'
                            'releases/{1}-{2}'.format(self.svn_path,
                                                      self.cp_name,
                                                      version),
                            '-m', 'Creating release branch'])
            self.log.info("Tagging release")
            check_call(['svn', 'copy',
                        '^{0}/branches/releases/{1}-{2}'.format(self.svn_path,
                                                                self.cp_name,
                                                                version),
                        '^{0}/tags/releases/{1}-{2}'.format(self.svn_path,
                                                            self.cp_name,
                                                            version),
                        '-m', 'Creating release branch'])
        else:
            self.rev = check_output('svnversion').strip()
            version = self.config.get('cp', 'content.pack.version')
            version = version.replace('-SNAPSHOT', '')
            cp_name = '{0}-cp-{1}-{2}-SNAPSHOT.jar'.format(self.cp_name,
                                                           version,
                                                           self.rev)
        os.mkdir('{0}/Lib'.format(self.path))
        self.log.info("Creating content pack {0}".format(cp_name))
        cp_path = "./target/{0}".format(cp_name)
        check_call(['jar', 'cf', cp_path, '-C', self.path, '.'])
        os.rmdir('{0}/Lib'.format(self.path))
        if release:
            version = self.increase_version(version)
            self.set_cp_version("{0}-SNAPSHOT".format(version))
            check_call(['svn', 'commit',
                        '-m', 'Updating contentpack.properties'])
        self.log.info("Done")

    def increase_version(self, version, increment=1):
        version = version.split('.')
        minor = int(version[-1])
        minor = minor + increment
        version[-1] = str(minor)
        ret = '.'.join(version)
        return ret

    def check_version_doesnt_exist(self, version):
        self.log.info("Checking release doesn't already exist in SVN")
        ret = call(['svn', 'info',
                    '^{0}/tags/releases/{1}-{2}'.format(self.svn_path,
                                                        self.cp_name,
                                                        version)])
        if ret == 0:
            raise errors.ReleaseAlreadyExists(version)
        return True

    def set_cp_version(self, version):
        self.log.info('Updating contentpack.properties version to'
                      ' {0}'.format(version))
        if self.pom_exists():
            self.set_pom_version(version)
        self.set_config_version(version)

    def read_props(self):
        prop_file = open('{0}/contentpack.properties'.format(self.path))
        config = StringIO.StringIO()
        config.write('[cp]\n')
        config.write(prop_file.read())
        config.seek(0, os.SEEK_SET)
        cp = ConfigParser.ConfigParser()
        cp.readfp(config)
        return cp

    def set_pom_version(self, version):
        pom_path = '{0}/pom.xml'.format(self.path)
        tree = ET.parse(pom_path)
        root = tree.getroot()
        cur_ver = root.find('./{http://maven.apache.org/POM/4.0.0}version')
        cur_ver.text = version
        tree.write(pom_path, xml_declaration=True,
                   encoding='utf-8', method="xml",
                   default_namespace='http://maven.apache.org/POM/4.0.0')

    def set_config_version(self, version):
        self.config.set('cp', 'content.pack.version', version)
        prop_file = '{0}/contentpack.properties'.format(self.path)
        with open(prop_file, 'w') as cf:
            self.config.write(cf)
        self.remove_whitespace_from_assignments()

    def pom_exists(self):
        return os.path.isfile('{0}/pom.xml'.format(self.path))

    def remove_whitespace_from_assignments(self):
        separator = "="
        config_path = '{0}/contentpack.properties'.format(self.path)
        lines = file(config_path).readlines()
        fp = open(config_path, "w")
        for line in lines:
            line = line.strip()
            if not line.startswith("#") and separator in line:
                assignment = line.split(separator, 1)
                assignment = map(str.strip, assignment)
                fp.write("%s%s%s\n" % (assignment[0],
                                       separator,
                                       assignment[1]))
            elif line.startswith('['):
                continue
            else:
                fp.write(line + "\n")


oo_client/hpoo.py
import requests
import json
import os.path
import oo_client.utils as utils
import oo_client.errors as errors
import logging
from uuid import UUID


class OORestCaller(object):
    def __init__(self, central_url, user, password, version="v1", ssl=True):
        # pylint: disable=fixme, line-too-long
        self.session = requests.Session()
        self.session.auth = (user, password)
        self.session.verify = ssl
        self.session.headers.update({'accept': 'application/json',
                                     'content-type': 'application/json'})
        self.central = central_url
        self.url = "{0}/oo/rest/{1}".format(central_url, version)
        self.log = logging.getLogger(self.__class__.__name__)

    def get(self, url_path, **kwargs):
        req = self.session.get("{0}/{1}".format(self.url, url_path),
                               params=kwargs)
        if req.status_code not in range(200, 205):
            raise errors.HTTPNon200(req.status_code, req.text)
        try:
            csrf = req.headers['X-CSRF-TOKEN']
            self.session.headers.update({'X-CSRF-TOKEN': csrf})
        except KeyError:
            # OO 10 added CSRF at some point version so not a problem
            # if we don't get a token
            pass
        if req.text:
            return json.loads(req.text)

    def put(self, url_path, data):
        req = self.session.put("{0}/{1}".format(self.url, url_path), data)
        if req.status_code not in range(200, 205):
            self.log.error(req.status_code)
            self.log.error(req.text)
            raise errors.HTTPNon200(req.status_code, req.text)
        if req.text:
            return json.loads(req.text)

    def post(self, url_path, data=None, files=None):
        if isinstance(data, dict):
            data = json.dumps(data)
        if files:
            self.session.headers.pop('content-type')
            req = self.session.post("{0}/{1}".format(self.url, url_path),
                                    files=files)
            self.session.headers.update({'content-type': 'application/json'})
        else:
            req = self.session.post("{0}/{1}".format(self.url, url_path), data)
        if req.status_code not in range(200, 205):
            self.log.error(req.status_code)
            self.log.error(req.text)
            raise errors.HTTPNon200(req.status_code, req.text)
        if req.text:
            return json.loads(req.text)


class OOClient(object):
    def __init__(self, central_url, user, password, version="v1", ssl=True):
        self.rest = OORestCaller(central_url, user, password, version=version,
                                 ssl=ssl)
        self.log = logging.getLogger(self.__class__.__name__)
        self.deploy_id = None
        # Checks OO is up and forces the CSRF token
        self.rest.get('version')
        # We have to get twice for valid CSRF because OO?
        self.rest.get('version')

    def get_new_deployment(self):
        return self.rest.post('deployments', None)['deploymentProcessId']

    def deploy_content_packs(self, path_list, timeout=3600):
        """
        Deploy a list of content packs
        return True if they all succeed, otherwise False
        """
        if not self.deploy_id:
            self.deploy_id = self.get_new_deployment()
            self.log.info("Got new deployment ID: {0}".format(self.deploy_id))
        for file_path in path_list:
            self.log.info("Starting uploads...")
            with open(file_path, 'rb') as cp:
                filename = os.path.basename(file_path)
                ret = self.rest.post("deployments/"
                                     "{0}/files".format(self.deploy_id),
                                     files={'file': (filename, cp)})
                self.log.info("Uploaded {0}".format(filename))
                self.log.debug(ret)
        ret = self.rest.put('deployments/{0}'.format(self.deploy_id), None)
        self.log.info("Deployment started, waiting up"
                      " to {0}s to complete".format(timeout))
        results = self.wait_for_deployment_to_complete(self.deploy_id,
                                                       timeout, 5)
        self.log.debug(results)
        all_res = []
        for cp, res in results['contentPackResponses'].items():
            api_response = res['responses'][0]['responseCategory']
            api_message = res['responses'][0]['message']
            if api_response != 'Success':
                self.log.error('{0}: {1}'.format(api_response, api_message))
                self.log.error("Failed to deploy {0}".format(cp))
                all_res.append(False)
            else:
                self.log.info("Deployed {0} successfully".format(cp))
                all_res.append(True)
        self.log.debug(all_res)
        if not all_res:
            return False
        return all(all_res)

    def deploy_content_pack(self, file_url_path):
        cp_name = os.path.splitext(os.path.basename(file_url_path))[0]
        with open(file_url_path) as cp:
            ret = self.rest.put("content-packs/{0}".format(cp_name), cp)
            cp_response = ret['contentPackResponses']
            api_responses = cp_response['{0}.jar'.format(cp_name)]['responses']
            api_response = api_responses[0]['responseCategory']
            api_message = api_responses[0]['message']
            if api_response != 'Success':
                self.log.error('{0}: {1}'.format(api_response, api_message))
                return False
            return True

    def is_deployment_complete(self, deploy_id):
        deployment = self.rest.get('deployments/{0}'.format(deploy_id))
        if deployment['status'] in ['FINISHED', 'FAILED']:
            return deployment['deploymentResultVO']
        else:
            return False

    def wait_for_deployment_to_complete(self, deploy_id, timeout, interval):
        func = utils.timeout(timeout, interval)(self.is_deployment_complete)
        return func(deploy_id)

    def run_flow_async(self, flow, run_name=None, inputs={}):
        if utils.validate_uuid4(flow):
            # this makes sure the UUID is formated how OO likes it
            uuid = str(UUID(flow))
        else:
            uuid = self.get_flow_uuid_from_path(flow)
        self.log.debug(uuid)
        payload = {'uuid': uuid,
                   'runName': run_name,
                   'inputs': inputs}
        self.log.info('Running flow: {0}'.format(flow))
        return self.rest.post('executions', data=payload)

    def run_flow(self, flow, run_name=None, inputs={}, timeout=300):
        '''
        Returns "RESOLVED", "ERROR", "DIAGNOSED", "NO_ACTION_TAKEN"
        or None if there is no result type.
        '''
        execution = self.run_flow_async(flow, run_name=run_name, inputs=inputs)
        run_id = execution['executionId']
        self.wait_for_run_to_complete(run_id, timeout, 5)
        self.log.info("Flow {0} FINISHED: {1}, link"
                      " {2}/oo/#/runtimeWorkspace/runs/"
                      "{3}".format(flow,
                                   self.get_run_result_type(run_id),
                                   self.rest.central, run_id))
        if self.get_run_status(run_id) == 'COMPLETED':
            return self.get_run_result_type(run_id)
        else:
            return None

    def run_flows(self, flows, timeout=300):
        ret = []
        for flow in flows:
            ret.append(self.run_flow(flow, timeout=timeout))
        if all([False if flow == 'ERROR' else True for flow in ret]):
            return True
        else:
            return False

    def is_run_complete(self, run_id):
        status = self.get_run_status(run_id)
        if status in ['COMPLETED', 'SYSTEM_FAILURE', 'CANCELED']:
            return True
        else:
            return False

    def wait_for_run_to_complete(self, run_id, timeout, interval):
        return utils.timeout(timeout, interval)(self.is_run_complete)(run_id)

    def get_flow_uuid_from_path(self, flow_path):
        folder_name = os.path.dirname(flow_path)
        flow_name = os.path.splitext(os.path.basename(flow_path))[0]
        flows = self.rest.get('flows/tree/level', path=folder_name)
        flow = [x['id'] for x in flows if x['name'] == flow_name]
        if len(flow) != 1:
            raise errors.NotFound('Flow not found with path'
                                  ' {0}'.format(flow_path))
        return flow[0]

    def get_run_status(self, run_id):
        return self.get_run_summary(run_id)['status']

    def get_run_result_type(self, run_id):
        return self.get_run_summary(run_id)['resultStatusType']

    def get_run_summary(self, run_id):
        summary = self.rest.get('executions/{0}/summary'.format(run_id))
        if len(summary) != 1:
            raise errors.NotFound('No run summary found for run id:'
                                  ' {0}'.format(run_id))
        return summary[0]

    def get_content_pack_id(self, name):
        cps = self.rest.get('content-packs')
        ret = [cp['id'] for cp in cps if cp['name'] == name]
        if len(ret) != 1:
            raise errors.NotFound('No content pack found with name'
                                  ' {0}'.format(name))
        return ret[0]

    def get_content_pack_from_flow(self, flow_path):
        uuid = self.get_flow_uuid_from_path(flow_path)
        return self.rest.get('flows/{0}'.format(uuid))['cpName']

    def get_all_flows_in_cp(self, cp_name):
        # pylint: disable=fixme, line-too-long
        cp_id = self.get_content_pack_id(cp_name)
        tree = self.rest.get('content-packs/{0}/content-tree'.format(cp_id))
        flows = {flow['id']: flow['path'] for flow in tree if flow['type'] == 'FLOW'}
        return flows

    # Add ability to operate configuration items
    def get_name_value_pair(self, output):
        name = output['name']
        val = output['value']
        ret = {'name': name, 'value': val}
        return ret

    def get_a_configuration_item(self, type, path):
        """
        Supported types: domain-terms, group-aliases, selection-lists,
        system-accounts, system-properties
        """
        ret = self.rest.get('config-items/{0}/{1}'.format(type, path))
        name_value = self.get_name_value_pair(ret)
        print json.dumps(name_value, indent=4)
        return name_value

    def set_a_configuration_item(self, type, path, data):
        # path includes name of the configuration item
        # data should be value
        if type == 'system-accounts':
            user_pass = data.partition(':')
            username = user_pass[0]
            password = user_pass[2]
            dct = {'username': username, 'password': password}
            data = json.dumps(dct)
        jdata = json.dumps(data)
        print jdata
        ret = self.rest.put('config-items/{0}/{1}'.format(type, path), jdata)
        res = self.get_name_value_pair(ret)
        print json.dumps(res, indent=4)
        return res

    def get_configuration_items_by_type(self, type):
        ret = self.rest.get('config-items/{0}'.format(type))
        name_val = [{'name': d['name'], 'value': d['value']} for d in ret]
        print json.dumps(name_val, indent=4)
        return name_val

    def get_all_configuration_items(self):
        ret = self.rest.get('config-items')
        print ret
        triple = [{'type': d['type'], 'name': d['name'], 'value': d['value']} for d in ret]
        print json.dumps(triple, indent=4)
        return triple
